11-cometbls/zk_verifier.go

```diff
@@ -0,0 +1,219 @@
+	if err != nil {
+		panic(fmt.Sprintf("could not read the verifying key: '%s'", vkHex))
+	}
+}
+
+func ParseZKP(data []byte) (*ZKP, error) {
+
+	zkp := ZKP{}
+
+	cursor := 0
+
+	zkp.Proof.A.SetBytes(data[0:G1_SIZE])
+	cursor += G1_SIZE
+
+	zkp.Proof.B.SetBytes(data[cursor : cursor+G2_SIZE])
+	cursor += G2_SIZE
+
+	zkp.Proof.C.SetBytes(data[cursor : cursor+G1_SIZE])
+	cursor += G1_SIZE
+
+	zkp.ProofCommitment.SetBytes(data[cursor : cursor+G1_SIZE])
+	cursor += G1_SIZE
+
+	zkp.ProofCommitmentPoK.SetBytes(data[cursor : cursor+G1_SIZE])
+
+	return &zkp, nil
+}
+
+func (zkp ZKP) Verify(trustedValidatorsHash []byte, header ProverLightHeader) error {
+	if len(header.ChainId) > 31 {
+		return errors.New("chain id length cannot be larger than 31")
+	}
+
+	commHash := commitmentsHash(zkp.ProofCommitment)
+	inpHash := inputsHash(header, trustedValidatorsHash)
+
+	var initialPoint curve.G1Affine
+	initialPoint.Add(&verifyingKey.G1.K[0], &zkp.ProofCommitment)
+
+	var commMul curve.G1Affine
+	var commBigInt big.Int
+	commHash.BigInt(&commBigInt)
+	commMul.ScalarMultiplication(&verifyingKey.G1.K[2], &commBigInt)
+
+	var inpMul curve.G1Affine
+	var inpBigInt big.Int
+	inpHash.BigInt(&inpBigInt)
+	inpMul.ScalarMultiplication(&verifyingKey.G1.K[1], &inpBigInt)
+
+	publicInputsMsm := initialPoint
+	publicInputsMsm.Add(&publicInputsMsm, &inpMul)
+	publicInputsMsm.Add(&publicInputsMsm, &commMul)
+
+	hasher := sha256.New()
+	a := zkp.Proof.A.Bytes()
+	hasher.Write(a[:])
+	c := zkp.Proof.C.Bytes()
+	hasher.Write(c[:])
+	msm := publicInputsMsm.Bytes()
+	hasher.Write(msm[:])
+
+	hasher.Reset()
+	pc := zkp.ProofCommitment.Bytes()
+	hasher.Write(pc[:])
+	pcPok := zkp.ProofCommitmentPoK.Bytes()
+	hasher.Write(pcPok[:])
+
+	alpha := verifyingKey.G1.Alpha
+	gamma := verifyingKey.G2.Gamma
+	delta := verifyingKey.G2.Delta
+	beta := verifyingKey.G2.Beta
+
+	// NOTE(aeryz): We didn't use fiat-shamir here since we don't batch two pairings. I'm not sure
+	// if we still need it though.
+	result, err := curve.PairingCheck([]curve.G1Affine{
+		zkp.Proof.A,
+		publicInputsMsm,
+		zkp.Proof.C,
+		alpha,
+	}, []curve.G2Affine{
+		zkp.Proof.B,
+		*gamma.Neg(&gamma),
+		*delta.Neg(&delta),
+		*beta.Neg(&beta),
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if !result {
+		return errors.New("proof verification failed")
+	}
+
+	return verifyingKey.CommitmentKey.Verify(zkp.ProofCommitment, zkp.ProofCommitmentPoK)
+}
+
+func hashToField(msg []byte) fr.Element {
+	hmac := hmac.New(Hash, []byte(CometblsHMACKey))
+	hmac.Write(msg)
+	modMinusOne := new(big.Int).Sub(fr.Modulus(), big.NewInt(1))
+	num := new(big.Int).SetBytes(hmac.Sum(nil))
+	num.Mod(num, modMinusOne)
+	num.Add(num, big.NewInt(1))
+	val, overflow := uint256.FromBig(num)
+	if overflow {
+		panic("impossible; qed;")
+	}
+	valBytes := val.Bytes32()
+	var element fr.Element
+	err := element.SetBytesCanonical(valBytes[:])
+	if err != nil {
+		panic("impossible; qed;")
+	}
+	return element
+}
```

Comment:
Cometbls is public you can reuse the function from our fork

Reply:
#2804